server/src/main/java/org/elasticsearch/action/search/SearchPhaseController.java

```diff
@@ -161,6 +161,7 @@ public static List<DfsKnnResults> mergeKnnResults(SearchRequest request, List<Df
 
         List<DfsKnnResults> mergedResults = new ArrayList<>(request.source().knnSearch().size());
         for (int i = 0; i < request.source().knnSearch().size(); i++) {
+            //TODO how do we merge via diversified top-k parents?
```

Comment:
I have a couple ideas here. My main thought is that kNNResults will not only include the child document ScoreDocs, but can also include an `int[]` of the matching parent documents. I would need some special logic here, but when combining shard level results, we can ensure that we combine the top `k` parent documents given the individual passage scores